source/common/stats/thread_local_store.cc

```diff
@@ -61,6 +61,22 @@ std::list<GaugeSharedPtr> ThreadLocalStoreImpl::gauges() const {
   return ret;
 }
 
+std::list<HistogramSharedPtr> ThreadLocalStoreImpl::histograms() const {
+  // Handle de-dup due to overlapping scopes.
+  std::list<HistogramSharedPtr> ret;
+  std::unordered_set<std::string> names;
+  std::unique_lock<std::mutex> lock(lock_);
+  for (ScopeImpl* scope : scopes_) {
+    for (auto histogram : scope->central_cache_.histograms_) {
+      if (names.insert(histogram.first).second) {
+        ret.push_back(histogram.second);
```

Comment:
probably worth a note here that the histograms will come back in arbitrary order.

suggestion: return a vector rather than a list, and build that vector after accumulating the set, so that:
  a) you can let go of the lock before converting to a vector
  b) you can do vector.reserve(names_.size()) before building it.

Reply:
@jmarantz have not addressed this yet. will address in next round. :-)

Reply:
I realize on re-read that there's an established pattern of using lists rather than for returning collections of stats.  @mattklein123 -- I'm curious why that choice?

But for this PR I retract this suggestion -- it probably doesn't save that much to release the lock earlier.  I would like to see some comments as to why this de-duping is necessary though; that seems like something must have gone wrong upstream for it to be needed.

Reply:
I don't remember why this was a list. I think this was historical. Free free to return a vector if you want to change the various usages. 

re: The de-dup, that is required due to the threaded store uses scopes.

Reply:
Shouldn't we incorporate the scopes into the histogram names somehow, so we have distinct ones for each scope?  It seems like just dropping them would lose data.

Reply:
Hmm yes, this is a problem for histograms which don't use shared mem backing store. It's not an issue for counters/gauges. Good catch. This will need to be fixed somehow.

Reply:
Cool... @ramaraochavali you can maybe leave a TODO here, and maybe even a warning log that we are dropping content?